tool lineage for versionless toolshed tool_ids

mvdbeek opened the pull request:
- 9b7be814b34494a4705cb19c44ce2039931f610f allows running/importing workflows and re-running tools from different toolsheds, where tools with the same owner and name, but different version are installed. (The functionality was introduced with #818, but in its current implementation requires that the same tool version is installed)
- 99cb275 fixes #2058

mvdbeek committed:
Fix reversed workflow upgrade message

mvdbeek committed:
Build tool lineage for versionless toolshed tool_ids

The tool_id for a toolshed tool contains the version as last element
(<toolshed>/<name>/<owner>/<id>/<version>). This means that looking up a tool
lineage using a tools' tool_id does not work if the tool is installed in a different
version than referenced by the tool_id. To allow this, we register the
tool_lineage also under a tool_id that is stripped of the version.

nsoranzo commented on lib/galaxy/tools/toolbox/lineages/factory.py:
```diff
@@ -2,6 +2,16 @@
+    if "/repos/" not in guid:
+        return None
+    tool_version = guid.split("/")[-1]
+    return guid.split(tool_version)[0]
```
This may be unsafe if tool_version appears somewhere else in `guid`, I think the following is better:

``` python
last_slash = guid.rfind('/')
return guid[:last_slash] # or guid[:last_slash + 1] if you need the trailing slash
```

mvdbeek replied:
good point, thanks @nsoranzo !

mvdbeek committed:
Remove version by slicing tool_id before last slash

As pointed out by @nsoranzo, this  may avoid conflicts when the version string
appears more than once in tool_id.

martenson commented:
Thanks @mvdbeek  and reviewers!
